src/lib/blkalloc/varsize_blk_allocator.cpp

```diff
@@ -405,251 +384,380 @@ void VarsizeBlkAllocator::fill_cache_in_portion(blk_num_t portion_num, blk_cache
-                BLKALLOC_LOG(TRACE, "Alloced first blk_num={}", s_alloc_resp.out_blks[0].to_string());
-
-                // Convert the response block cache entries to blkids
-                blk_cache_entries_to_blkids(s_alloc_resp.out_blks, out_blkids);
-                total_allocated = s_alloc_resp.nblks_alloced;
-                break;
-            } else {
-                discard_current_allocation();
-                if ((retry + 1) < max_retries) {
-                    COUNTER_INCREMENT(m_metrics, num_retries, 1);
-                    auto const min_nblks = std::max< blk_count_t >(m_cfg.highest_slab_blks_count() * 2, nblks);
-                    BLKALLOC_LOG(
-                        DEBUG,
-                        "Failed to allocate {} blks from blk cache, requesting refill at least {} blks and retry={}",
-                        nblks, min_nblks, retry);
-                    request_more_blks_wait(nullptr /* seg */, min_nblks);
-                }
-            }
-        }
-        free_excess_blocks();
+    if (!hints.is_contiguous && !out_blkid.is_multi()) {
+        HS_DBG_ASSERT(false, "Invalid Input: Non contiguous allocation needs MultiBlkId to store");
+        return BlkAllocStatus::INVALID_INPUT;
     }
 
-    if (hints.is_contiguous) {
-        // failed to allocate in slab try direct.
-        if (status != BlkAllocStatus::SUCCESS) {
-            blk_count_t num_allocated{0};
-            status = alloc_blks_direct(nblks, hints, out_blkids, num_allocated);
-            if (status == BlkAllocStatus::SUCCESS) {
-                total_allocated += num_allocated;
-                BLKALLOC_LOG(TRACE, "Alloced blk_num={} directly", out_blkids.back().to_string());
-            }
-        }
-    } else {
-        if (status != BlkAllocStatus::SUCCESS) {
-            // try to allocate remainder
-            const blk_count_t nblks_remaining = static_cast< blk_count_t >(nblks - total_allocated);
-            BLKALLOC_LOG(DEBUG, "nblks={} failed to alloc all from fb cache, trying to alloc rest from bitset directly",
-                         nblks_remaining);
-            blk_count_t num_allocated{0};
-            auto status2 = alloc_blks_direct(nblks_remaining, hints, out_blkids, num_allocated);
-            if ((status2 == BlkAllocStatus::SUCCESS) || (status2 == BlkAllocStatus::PARTIAL)) {
-                total_allocated += num_allocated;
-                BLKALLOC_LOG(TRACE, "Alloced additional blk_num={} directly", out_blkids.back().to_string());
-            } else {
-                // failure to get more is really partial if we have some
-                BLKALLOC_LOG(TRACE, "Failed to alloc additional blks directly with code {}", status2);
-                if (status == BlkAllocStatus::PARTIAL) status2 = BlkAllocStatus::PARTIAL;
-            }
-            status = status2;
+    MultiBlkId tmp_blkid;
+    MultiBlkId& out_mbid = out_blkid.is_multi() ? r_cast< MultiBlkId& >(out_blkid) : tmp_blkid;
```

Comment:
can we fallback to alloc_contiguous if `is_multi() = false` ?

Or maybe instead of having `alloc_contiguous` and `alloc`,  we have overload `alloc` 
```
  blk_count_t nblks, blk_alloc_hints const& hints, MultiBlkId& out_blkid)  ->original alloc
  blk_count_t nblks, blk_alloc_hints const& hints, BlkId& out_blkid).  --> alloc_contiguous
 ```